# 1. Extract Data

In [2]:
# Thư viện cần thiết
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sqlalchemy import create_engine

In [3]:
# Kết nối MySQL

engine = create_engine("mysql+pymysql://root:26052004@localhost:3306/movie_ticket")

In [4]:
# Đọc dữ liệu từ các bảng trong MySQL

df_campaign = pd.read_sql("SELECT * FROM campaign", con=engine)
df_customer = pd.read_sql("SELECT * FROM customer", con=engine)
df_device = pd.read_sql("SELECT * FROM device", con=engine)
df_status = pd.read_sql("SELECT * FROM status", con=engine)
df_ticket = pd.read_sql("SELECT * FROM ticket", con=engine)

# 2. Cleaning Data
Làm sạch dữ liệu trước khi phân tích, trước tiên kiểm tra và đảm bảo 3 tiêu chí trước:
- Đúng kiểu dữ liệu (Data types) của từng cột
- Không bị Null
- Không trùng lặp (Duplicates) 

## 2.1. Bảng Campaign

In [5]:
df_campaign

campaign_id    campaign_type
0         106460  direct discount
1          30040  direct discount
2          13810          voucher
3          78370          voucher
4          83700          voucher
..           ...              ...
211       104140  direct discount
212        17580  direct discount
213       104980          voucher
214        12240          voucher
215        73330          voucher

[216 rows x 2 columns]

In [6]:
df_campaign.info()
df_campaign['campaign_id'].duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 216 entries, 0 to 215
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   campaign_id    216 non-null    int64 
 1   campaign_type  216 non-null    object
dtypes: int64(1), object(1)
memory usage: 3.5+ KB


np.int64(0)

## 2.2. Bảng Customer

In [7]:
df_customer

customer_id  usergender         dob
0            100032      Female    8/8/1985
1            100046        Male   7/11/1987
2            100050        Male  11/19/1994
3            100063        Male   9/10/1999
4            100076      Female   5/14/1995
...             ...         ...         ...
131395       231310  Not verify    1/1/1970
131396       231335  Not verify    1/1/1970
131397       231382  Not verify    1/1/1970
131398       231388  Not verify    1/1/1970
131399       231390  Not verify    1/1/1970

[131400 rows x 3 columns]

In [8]:
df_customer.info()
df_customer['customer_id'].duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131400 entries, 0 to 131399
Data columns (total 3 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   customer_id  131400 non-null  int64 
 1   usergender   131400 non-null  object
 2   dob          131400 non-null  object
dtypes: int64(1), object(2)
memory usage: 3.0+ MB


np.int64(0)

In [9]:
# Chuyển cột dob sang datetime
from datetime import datetime
df_customer['dob'] = pd.to_datetime(df_customer['dob'], errors='coerce')

In [10]:
df_customer['dob'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 131400 entries, 0 to 131399
Series name: dob
Non-Null Count   Dtype         
--------------   -----         
131400 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 1.0 MB


## 2.3. Bảng Device

In [11]:
df_device

device_number                model platform
0       00006afbe30ae7018c92bb324cd58afc              browser  website
1       0000b0ce524ef4d66c7bfdad67a91970          devicemodel   mobile
2       0000e2a055b67fe49c06bbe28b268251           iPhone10,2   mobile
3       00012f7a6842cca5d531b7c8b32967d6           iPhone10,3   mobile
4       000215f1b9363b5465b8ee895fa6c920          devicemodel   mobile
...                                  ...                  ...      ...
139897  fffc3c61f1caf84318398963a466a9c5          devicemodel   mobile
139898  fffd7948588c43fa27edccf9878bc994  Xiaomi Redmi Note 3   mobile
139899  fffedb35c487fc2e46dfc811c96f0aa8     Samsung SM-A530F   mobile
139900  ffffff09758e9f5f9935b5ebc8c0173e           iPhone10,3   mobile
139901                              None              browser  website

[139902 rows x 3 columns]

In [12]:
df_device.info()
df_device.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 139902 entries, 0 to 139901
Data columns (total 3 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   device_number  139901 non-null  object
 1   model          132763 non-null  object
 2   platform       139902 non-null  object
dtypes: object(3)
memory usage: 3.2+ MB


np.int64(0)

In [13]:
df_device.isnull().sum()

device_number       1
model            7139
platform            0
dtype: int64

In [14]:
# Thay giá trị null trong các cột model và device_number
df_device['model'].fillna('unknown', inplace=True)
df_device['device_number'].fillna('unknown', inplace=True)

C:\Users\levan\AppData\Local\Temp\ipykernel_21924\4164148636.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_device['model'].fillna('unknown', inplace=True)
C:\Users\levan\AppData\Local\Temp\ipykernel_21924\4164148636.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For e

In [15]:
df_device.isnull().sum()

device_number    0
model            0
platform         0
dtype: int64

## 2.4. Bảng Status

In [16]:
df_status

status_id                                        description error_group
0          1                                   Order successful        None
1         -1                                    Payment overdue    customer
2         -2  Insufficient funds in customer account. Please...    customer
3         -3                         No response from your bank    external
4         -4  Password locked due to multiple incorrect atte...    customer
5         -5                           Payment failed from bank    external
6         -6               Need verify your account to continue    customer
7         -7                    Transaction temporarily limited    internal

In [17]:
df_status.info()
df_status.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   status_id    8 non-null      int64 
 1   description  8 non-null      object
 2   error_group  7 non-null      object
dtypes: int64(1), object(2)
memory usage: 324.0+ bytes


status_id      0
description    0
error_group    1
dtype: int64

In [18]:
# fill null 
df_status['error_group'].fillna('success', inplace=True)

C:\Users\levan\AppData\Local\Temp\ipykernel_21924\347458481.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_status['error_group'].fillna('success', inplace=True)


## 2.5. Bảng ticket

In [19]:
df_ticket

ticket_id  customer_id paying_method  \
0       4f5200dcdcf2396b8d50ff84bf423f32       100009  money in app   
1       07abbaf28c772692f0367ad992bb3184       100493  bank account   
2       1fdbeb7eceba8a27d9d985b5b70c219a       100596  money in app   
3       7715c9955866bd296b98543412839abd       100852  bank account   
4       776efd7bda0b715084430e6385f67746       100965  money in app   
...                                  ...          ...           ...   
154822  6d81d62a2d26adb4c9b47d254a2bfe27       194726    debit card   
154823  740855b0fdc18125b773f904279d2dfb       211166  money in app   
154824  9a65bf4e0ab9da0186007404f6ec0960       211671  money in app   
154825  e47b196dbb5b487b3f6cf1cabe242552       223500  bank account   
154826  809798c0b269b8090cb658a8f9c2d13f       223935  money in app   

        theater_name                     device_number  original_price  \
0               13.0  244764a57dbdeb8fe9b164847ad55183            9.90   
1              180.0  8fa83cf46284aafd6e5da6c96f7862b5            8.66   
2               56.0  69cf9244654949047f006e441fa7a8a7           10.31   
3              119.0  879ed11af9d6d2b5cda4d299590735a7           27.75   
4              107.0  0a233600d993a02cc1d39fb0d87fc7de            4.33   
...              ...                               ...             ...   
154822          70.0  f5ad6683cca252327b19bef18f0e8a65            5.16   
154823         101.0  217e197a017719ed96da9c140968a15f            5.07   
154824         176.0  627e3c4a56079dff9fd3a6f0f77ae13c           14.72   
154825         176.0  f6836889dc1bbc02547e23a1c59a8bb2           13.36   
154826           6.0  f06e6c483fe67adee4c8a7410e21d787           10.31   

        discount_value  final_price                     time  status_id  \
0                 2.10         7.80  2022-07-08 17:46:36.145          1   
1                 1.48         7.18  2022-07-15 20:44:09.952          1   
2                 2.31         8.00  2022-07-30 11:16:55.483          1   
3                 1.03        26.72  2022-07-04 22:33:07.458          1   
4                 1.03         3.30  2022-07-03 18:41:45.098          1   
...                ...          ...                      ...        ...   
154822            2.31         2.85  2022-10-22 13:01:58.840         -5   
154823            0.00         5.07  2022-10-23 10:51:32.185          1   
154824            2.31        12.41  2022-10-23 11:11:23.037          1   
154825            4.78         8.58  2022-10-22 20:35:02.732          1   
154826            2.31         8.00  2022-10-22 16:00:20.072          1   

        campaign_id              movie_name  
0             83330  Thor: Love And Thunder  
1             83330  Thor: Love And Thunder  
2                 0  Thor: Love And Thunder  
3             85940  Thor: Love And Thunder  
4             85940  Thor: Love And Thunder  
...             ...                     ...  
154822        97120              Black Adam  
154823            0              Black Adam  
154824        97120              Black Adam  
154825        97150              Black Adam  
154826        97120              Black Adam  

[154827 rows x 12 columns]

In [20]:
df_ticket.info()
df_ticket.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154827 entries, 0 to 154826
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   ticket_id       154827 non-null  object 
 1   customer_id     154827 non-null  int64  
 2   paying_method   154827 non-null  object 
 3   theater_name    154827 non-null  float64
 4   device_number   154827 non-null  object 
 5   original_price  154827 non-null  float64
 6   discount_value  154827 non-null  float64
 7   final_price     154827 non-null  float64
 8   time            154827 non-null  object 
 9   status_id       154827 non-null  int64  
 10  campaign_id     154827 non-null  int64  
 11  movie_name      154827 non-null  object 
dtypes: float64(4), int64(3), object(5)
memory usage: 14.2+ MB


np.int64(102)

In [21]:
# Chuyển cột time sang datetime
df_ticket['time'] = pd.to_datetime(df_ticket['time'], errors='coerce')

In [22]:
df_ticket.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154827 entries, 0 to 154826
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   ticket_id       154827 non-null  object        
 1   customer_id     154827 non-null  int64         
 2   paying_method   154827 non-null  object        
 3   theater_name    154827 non-null  float64       
 4   device_number   154827 non-null  object        
 5   original_price  154827 non-null  float64       
 6   discount_value  154827 non-null  float64       
 7   final_price     154827 non-null  float64       
 8   time            154827 non-null  datetime64[ns]
 9   status_id       154827 non-null  int64         
 10  campaign_id     154827 non-null  int64         
 11  movie_name      154827 non-null  object        
dtypes: datetime64[ns](1), float64(4), int64(3), object(4)
memory usage: 14.2+ MB


In [23]:
df_ticket.drop_duplicates(inplace=True)

## 2.6. Merge các bảng vào với bảng ticket

In [24]:
# bang ticket la bang chinh, ket hop voi cac bang con lai

df_join_campaign = pd.merge(df_ticket, df_campaign, on='campaign_id', how='left')
df_join_customer = pd.merge(df_join_campaign, df_customer, on='customer_id', how='left')
df_join_device = pd.merge(df_join_customer, df_device, on='device_number', how='left')
df_final = pd.merge(df_join_device, df_status, on='status_id', how='left')

In [25]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 154725 entries, 0 to 154724
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   ticket_id       154725 non-null  object        
 1   customer_id     154725 non-null  int64         
 2   paying_method   154725 non-null  object        
 3   theater_name    154725 non-null  float64       
 4   device_number   154725 non-null  object        
 5   original_price  154725 non-null  float64       
 6   discount_value  154725 non-null  float64       
 7   final_price     154725 non-null  float64       
 8   time            154725 non-null  datetime64[ns]
 9   status_id       154725 non-null  int64         
 10  campaign_id     154725 non-null  int64         
 11  movie_name      154725 non-null  object        
 12  campaign_type   91627 non-null   object        
 13  usergender      154725 non-null  object        
 14  dob             154725 non-null  dat

In [26]:
df_final.isnull().sum()

ticket_id             0
customer_id           0
paying_method         0
theater_name          0
device_number         0
original_price        0
discount_value        0
final_price           0
time                  0
status_id             0
campaign_id           0
movie_name            0
campaign_type     63098
usergender            0
dob                   0
model                78
platform             78
description           0
error_group           0
dtype: int64

In [27]:
# Khac phuc loi null
# - campaign_type: Thay the bang 'not available' => neu khong co thong tin ve campaign_type thi co the la khong co campaign nao ho tro
# - model: Thay the bang 'browser' => neu khong co thong tin ve model thi co the la truy cap bang trinh duyet web
# - platform: Thay the bang 'website' => neu khong co thong tin ve platform thi co the la truy cap bang website

df_final['campaign_type'].fillna('not available', inplace=True)
df_final['model'].fillna('browser', inplace=True)
df_final['platform'].fillna('website', inplace=True)

C:\Users\levan\AppData\Local\Temp\ipykernel_21924\3986595765.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_final['campaign_type'].fillna('not available', inplace=True)
C:\Users\levan\AppData\Local\Temp\ipykernel_21924\3986595765.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a

In [28]:
# cac gia tri khac nhau trong moi cot
df_final.nunique()

ticket_id         154725
customer_id       119477
paying_method          5
theater_name         179
device_number     126459
original_price      1895
discount_value       242
final_price         2715
time              154725
status_id              8
campaign_id          211
movie_name           253
campaign_type          4
usergender             3
dob                11322
model               1215
platform               2
description            8
error_group            4
dtype: int64

Nhận xét từ bộ dữ liệu
| **Cột**           | **Tên**                     | **Nhận xét / Số Giá trị khác nhau**                                                      |
|--------------------|-----------------------------|-----------------------------------------------------------------------------------------|
| **ticket_id**      | Mã vé                       | 154725                                                                                  |
| **customer_id**    | Mã khách hàng               | 119477 → Có nhiều khách hàng mua nhiều lần                                              |
| **paying_method**  | Phương thức thanh toán       | 5 → money in app, bank account, credit card, debit card, other                          |
| **theater_name**   | Tên rạp                     | 179                                                                                     |
| **device_number**  | ID thiết bị                 | 126459 → lớn hơn customer_id nhưng nhỏ hơn ticket_id. Một KH có thể mua nhiều lần, nhiều thiết bị |
| **original_price** | Giá gốc                     | 1895                                                                                    |
| **discount_value** | Giá trị giảm giá            | 242                                                                                     |
| **final_price**    | Giá cuối (sau giảm giá)     | 2715                                                                                    |
| **time**           | Thời gian đặt vé            | 154725                                                                                  |
| **status_id**      | Mã trạng thái/lỗi           | 8                                                                                       |
| **campaign_id**    | Mã chiến dịch               | 211                                                                                     |
| **movie_name**     | Tên phim                    | 253                                                                                     |
| **campaign_type**  | Kiểu chiến dịch             | 4 → direct discount, not available, voucher, reward point                               |
| **usergender**     | Giới tính                   | 3 → Female, Male, Not verify                                                            |
| **dob**            | Ngày sinh khách hàng        | 11322                                                                                   |
| **model**          | Model thiết bị              | 1215                                                                                    |
| **platform**       | Nền tảng                    | 2 → mobile, website                                                                     |
| **description**    | Mô tả lỗi                   | 8                                                                                       |
| **error_group**    | Nhóm lỗi                    | 4 → success, external, customer, internal                                               |


## 2.7. Load ngược vào MySQL

In [29]:
df_final.to_sql(
    name="ticket_final",   # Tên bảng muốn đặt
    con=engine,
    if_exists="replace",   # 'replace' ghi đè, 'append' ghi thêm
    index=False
)

154725